File: src/main/java/tiagoanacleto/bookstore/fileprocessor/impl/JsonFileProcessor.java
package tiagoanacleto.bookstore.fileprocessor.impl;

import lombok.extern.log4j.Log4j;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;
import tiagoanacleto.bookstore.fileprocessor.FileProcessor;
import tiagoanacleto.bookstore.model.Book;

import java.io.FileReader;
import java.io.IOException;
import java.io.Reader;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * This class implements the FileProcessor interface
 * with json file type processing behavior
 */
@Log4j
public class JsonFileProcessor implements FileProcessor {

    @Override
    public Map<Long, Book> processFile(String path) {
        JSONParser jsonParser = new JSONParser();
        Map books = new HashMap<>();

        try (Reader fileReader = new FileReader(path)) {
            JSONArray lines = (JSONArray) jsonParser.parse(fileReader);

            books = (Map<Long, Book>) lines.stream()
                                .collect(Collectors.toMap(
                                            c -> (Long) c.get("ISBN"), mapToBookObj));

        } catch (IOException | ParseException e) {
            log.error(e.getMessage());
        }

        return books;
    }

    private Function<JSONObject, Book> mapToBookObj = json -> {
        Book book = new Book();
        book.setTitle((String) json.get("Title"));
        book.setDescription((String) json.get("Description"));
        book.setIsbn((Long) json.get("ISBN"));
        book.setAuthor((String) json.get("Author"));
        book.setGenre((String) json.get("Genre"));
        book.setPages(String.valueOf((Long) json.get("Pages")));
        book.setAgeRange((String) json.get("Age Range"));
        book.setPrice((String) json.get("Price"));
        book.setQty((Long) json.get("Qty."));

        return book;
    };
}

File: src/main/kotlin/bookstore/fileprocessor/impl/JsonFileProcessor.kt
package bookstore.fileprocessor.impl

import bookstore.fileprocessor.FileProcessor
import bookstore.model.Book
import org.json.simple.JSONArray
import org.json.simple.JSONObject
import org.json.simple.parser.JSONParser
import org.json.simple.parser.ParseException
import java.io.FileReader
import java.io.IOException
import java.util.stream.Collectors

/**
 * This class implements the FileProcessor interface
 * with json file type processing behavior
 */
class JsonFileProcessor : FileProcessor {

    override fun processFile(path: String): Map<Long, Book> {
        val jsonParser = JSONParser()
        var books: Map<Long, Book> = HashMap()

        try {
            FileReader(path).use { fileReader ->
                val lines = jsonParser.parse(fileReader) as JSONArray
                books = lines.stream()
                            .collect(Collectors.toMap({
                                    c: JSONObject -> c.get("ISBN")
                                }, mapToBookObj)
                            ) as Map<Long, Book>
                }
        } catch (e: IOException) {
            println(e.message);
        } catch (e: ParseException) {
            println(e.message);
        }
        return books
    }

    private val mapToBookObj = {
            json: JSONObject? ->
                val book = Book(json!!["ISBN"] as Long)
                book.title = json["Title"] as String
                book.description = json["Description"] as String
                book.author = json["Author"] as String
                book.genre = json["Genre"] as String
                book.pages = (json["Pages"] as Long?).toString()
                book.ageRange = json["Age Range"] as String
                book.price = json["Price"] as String
                book.qty = json["Qty."] as Long
                book
            }
}